src/test/java/com/commercetools/sync/producttypes/utils/ProductTypeReferenceReplacementUtilsTest.java

```diff
@@ -0,0 +1,347 @@
+                + "'nestedattr'. Cause: ProductType reference is not expanded.");
+    }
+
+    @Test
+    void replaceProductTypesReferenceIdsWithKeys_WithNestedTypeWithSetOfSetOfSet_ShouldReplaceRef() {
+        // preparation
+        final ProductType referencedProductType = mock(ProductType.class);
+        when(referencedProductType.getKey()).thenReturn("referencedProductType");
+
+        final Reference<ProductType> productTypeReference = spy(ProductType.reference(referencedProductType));
+        when(productTypeReference.getObj()).thenReturn(referencedProductType);
+
+        final AttributeDefinition nestedTypeAttr = AttributeDefinitionBuilder
+            .of("nestedattr", ofEnglish("nestedattr"),
+                SetAttributeType.of(
+                    SetAttributeType.of(
+                        SetAttributeType.of(
+                            NestedAttributeType.of(productTypeReference)))))
+            .build();
+
+        final ProductType productType = mock(ProductType.class);
+        when(productType.getKey()).thenReturn("withNestedTypeAttr");
+        when(productType.getAttributes()).thenReturn(singletonList(nestedTypeAttr));
+
+        final List<ProductType> productTypes = singletonList(productType);
+
+        // test
+        final List<ProductTypeDraft> productTypeDrafts = replaceProductTypesReferenceIdsWithKeys(productTypes);
+
+        // assertion
+        assertThat(productTypeDrafts)
+            .hasOnlyOneElementSatisfying(productTypeDraft -> {
+                final SetAttributeType setAttributeType = (SetAttributeType)
+                    productTypeDraft.getAttributes().get(0).getAttributeType();
+                final SetAttributeType setOfSet = (SetAttributeType) setAttributeType.getElementType();
+                final SetAttributeType setOfSetOfSet = (SetAttributeType) setOfSet.getElementType();
+                final NestedAttributeType nestedAttributeType = (NestedAttributeType) setOfSetOfSet.getElementType();
+                assertThat(nestedAttributeType.getTypeReference().getId()).isEqualTo(referencedProductType.getKey());
+            });
+    }
+
+    @Test
+    void buildProductTypeQuery_WithNoParam_ShouldReturnQueryWithAllNeededReferencesExpanded() {
+        final ProductTypeQuery productTypeQuery = buildProductTypeQuery();
+        assertThat(productTypeQuery.expansionPaths()).containsExactly(
+            ExpansionPath.of("attributes[*].type.typeReference"));
+    }
+
+    @Test
+    void buildProductTypeQuery_With0MaxSetDepth_ShouldReturnQueryWithAllNeededReferencesExpanded() {
+        final ProductTypeQuery productTypeQuery = buildProductTypeQuery(0);
+        assertThat(productTypeQuery.expansionPaths()).containsExactly(
+            ExpansionPath.of("attributes[*].type.typeReference"));
+    }
+
+    @Test
+    void buildProductTypeQuery_With1MaxSetDepth_ShouldReturnQueryWithAllNeededReferencesExpanded() {
```

Comment:
it looks like you forgot to add `With2MaxSetDepth` test 😄

Reply:
f4d31fa